packages/api/src/store/db.ts

```diff
@@ -195,17 +205,19 @@ export class DB {
     query: string | QueryConfig<I>,
     values?: I
   ): Promise<QueryResult<T>> {
-    let queryLog;
+    let queryLog: string;
     if (typeof query === "string") {
       queryLog = JSON.stringify({ query: query.trim(), values });
     } else {
       queryLog = JSON.stringify(query);
     }
-    let result;
+    let result: QueryResult;
     logger.info(`runQuery phase=start query=${queryLog}`);
     const start = Date.now();
+    const queryTimer = this.metricHistogram.startTimer();
```

Comment:
I think the most useful part of this metric would be to have it broke down by the actual query being executed. 

In the ideal version, we would specify the table + the function that is being executed, for example `StreamTable#getByStreamKey` or something like that. There **might** be a way to grab the "caller" information automatically here, but I'm not so sure myself how to do it. On C# there is an attribute called [`[CallerMemberName]`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callermembernameattribute?view=net-6.0) which was filled by the compiler and could be used for that, but I don't think there's such a thing in TypeScript.

So I see a couple of options now, which we can discuss further on the most appropriate (or come up/decide for some other):
 1. Stick to trying to determine the caller automatically somehow. One idea is to use the stack trace for that, but the implementation could get quite messy and hard to maintain, especially with all the `async` stuff going on and the common `Table` base class that we have for all tables (stack trace might show that instead of the "real" table?). May also get less useful when different queries can be made by the same function, like for example from [`filter`](https://github.com/livepeer/livepeer-com/blob/049c0fa1442db71201f90ea5eed49e627e69973c/packages/api/src/store/table.ts#L114) function, which accepts a list of filters and builds a dynamic query from them, so they can vary in performance **a lot** (we just had one set of filters there [which wasn't using any index at all](https://github.com/livepeer/livepeer-com/pull/793)).
 2. Set a label on the metric with the full `query` but without the `values` (what to do when `typeof query !== "string"`?). I think that will be the easiest to implement, but might have issues with the length of the prometheus labels? Not sure, but maybe it can take some labels with 100s of characters with no issues. If you could make sure that is the case, I'd be down with just doing this then. Just gotta figure out when does that `typeof query !== "string"` happen and what do do about it.
 3. Have some new parameter(s) here like `queryName` or `queryId` (or separate `table` and `queryName`), which is passed explicitly by every caller. This seems to be the most bulletproof, so I'm growing more fond of it, it might just be a little cumbersome to implement initially. With it, we could also circumvent all issues mentioned above, by sometimes building the `queryId` dynamically in the case of the dynamic `find` query for example (tho on those cases it might as well grow very similar to just labeling with the whole `query`).

Reply:
I was thinking for splitting the `queryLog` string and selecting the first element (to identify CRUD operation), but for fetching the table information, either the function signature would need change or we introduce a sql parser (not recommended imo).

setting a label with the full query would split the metric across multiple dimensions and might make it more difficult to aggregate results (although, i'm unfamiliar with prometheus in this regard).

I'm also in favour of providing the table name as a parameter. Labels can be dynamically updated for the histogram once the `startTimer()` call has been issued. The implementation might pose a bigger challenge as you stated, since all `.query` and `.replicaQuery` calls would need to be updated (or we introduce overloaded signatures for these functions?).